File: src/game/tetfield.c
#include "tetmino.h"
#include "tetgrid.h"
#include "tetfield.h"

static void stop_retrying_action(struct tetfield *tf)
{
	if (tf->last_action == HARDDROP) {
		tf->last_action = NO_ACTION;
		tf->timeout[RETRY_ACTION] = 0;
		tf->timeout[HARDDROP] += SPAWN_DELAY - AUTOREPEAT_FRAMES;
	}
	if (tf->last_action != NO_ACTION) {
		tf->last_action = NO_ACTION;
		tf->timeout[RETRY_ACTION] += SPAWN_DELAY - AUTOREPEAT_FRAMES;
	}
}

void enter_tetfield(struct tetfield *tf, int piece, int col)
{
	int delay = SPAWN_DELAY;
	if (delay < tf->gravity)
		delay = tf->gravity;
	init_tetmino(&tf->mino, piece, SPAWN_ROW + 1, col, delay + 1);
	tf->state = TETFIELD_SPAWN;
	/* prevent last action from cancelling spawn delay */
	stop_retrying_action(tf);
}

static int is_retrying_action(struct tetfield *tf)
{
	enum action a = tf->last_action;
	return a != NO_ACTION && tf->timeout[RETRY_ACTION] > tf->timeout[a];
}

static enum action make_move(struct tetfield *tf, enum action a)
{
	if (tf->timeout[a] == 0 && control_tetmino(&tf->mino, tf->grid->blocks, a)) {
		if (!is_retrying_action(tf) || tf->timeout[tf->last_action] == 0) {
			tf->last_action = a;
		}
		tf->timeout[RETRY_ACTION] = AUTOREPEAT_FRAMES;
		tf->timeout[a] = AUTOREPEAT_FRAMES;
		return a;
	}
	return NO_ACTION;
}

static void dec_timeout(struct tetfield *tf)
{
	int i;
	for (i=0; i < END_ACTION; i++) {
		if (tf->timeout[i])
			tf->timeout[i]--;
	}
}

/* process moves before the piece is spawned */
static void update_prespawn(struct tetfield *tf, enum action a)
{
	if (!tf->timeout[RETRY_ACTION]) {
		switch (a) {
		case ROTATE_CW:
		case ROTATE_CCW:
			tf->last_action = a;
			break;
		case MOVE_RIGHT:
		case MOVE_LEFT:
		case HARDDROP:
		case SOFTDROP:
			if (!tf->timeout[a]) {
				tf->mino.falling = 1;
				tf->timeout[a] = 1;
			}
			break;
		default:
			break;
		}
	}
}

static enum action update_move(struct tetfield *tf, enum action a)
{
	enum action moved = NO_ACTION;

	if (tf->timeout[a]) {
		if (tf->timeout[RETRY_ACTION]) {
			moved = make_move(tf, tf->last_action);
		}
		if (a != NO_ACTION) {
			tf->last_action = a;
			tf->timeout[RETRY_ACTION] = AUTOREPEAT_FRAMES;
		}
	} else if ((moved = make_move(tf, a)) == NO_ACTION) {
		if (a != NO_ACTION) {
			tf->last_action = a;
			tf->timeout[RETRY_ACTION] = WALL_CHARGE_FRAMES;
		}
	}

	return moved;
}

static int is_movable(const struct tetmino *t)
{
	return t->falling || (t->lock_delay_move && t->lock_delay_step);
}

/* check that there are no unfinished line clears at occupied rows */
static int is_lockable(const struct tetfield *tf)
{
	const blocks_row *blocks = tf->grid->blocks + tf->mino.row;
	int i;
	for (i=0; i < PIECE_HEIGHT; i++) {
		if (!(blocks[i] & LINE_CLEAR_MARK) && tetmino_has_row(tf->mino.shape, i))
			return 0;
	}
	return 1;
}

int run_tetfield(struct tetfield *tf, enum action a, struct changed *out)
{
	struct tetgrid *grid = tf->grid;

	dec_timeout(tf);
	out->moved = NO_ACTION;
	out->dropped = 0;

	switch (tf->state) {
	case TETFIELD_SPAWN:
		update_prespawn(tf, a);
		if (tf->mino.falling > 1 && grid->clearing) {
			/* stop spawn timer during line clear animation */
			tf->mino.falling++;
		}
		out->dropped = update_tetmino(&tf->mino, grid->blocks, tf->gravity);
		if (!out->dropped) {
			if (is_movable(&tf->mino) || grid->clearing)
				return 1;
			tf->state = TETFIELD_TOP_OUT;
			return 0;
		}
		tf->timeout[RETRY_ACTION] = 1;
		tf->state = TETFIELD_MOVE;
		out->moved = update_move(tf, a);
		break;
	case TETFIELD_MOVE:
		out->moved = update_move(tf, a);
		out->dropped = update_tetmino(&tf->mino, grid->blocks, tf->gravity);
		break;
	case TETFIELD_PLACED:
		/* accepting no more actions */
		out->dropped = control_tetmino(&tf->mino, grid->blocks, HARDDROP);
		break;
	case TETFIELD_TOP_OUT:
		return 0;
	}

	/* set timer to block actions for next tetromino during spawn delay */
	if (tf->timeout[RETRY_ACTION] == 1) {
		stop_retrying_action(tf);
	}

	update_removable_rows(grid, &tf->mino);

	/* return 0 on lock condition */
	return is_movable(&tf->mino) || (tf->state = TETFIELD_PLACED, !is_lockable(tf));
}

static void levelup(struct tetfield *tf)
{
	if (tf->gravity > 1) {
		tf->gravity *= 0.9;
		tf->pieces = 0;
	}
}

static int get_score_for_lines(struct tetfield *tf, int num_lines_cleared)
{
	int score = tf->grid->cols * num_lines_cleared;

	while (num_lines_cleared > 2) {
		score *= 2;
		num_lines_cleared -= 2;
	}

	if (num_lines_cleared > 1) {
		score *= 1.5;
	}

	int gravity = tf->gravity;

	int level = (30 - gravity) * 10 / gravity;

	if (gravity < 5) {
		level = (10 - gravity + tf->pieces / 25) * 10;
	}

	if (level > 0) {
		score *= level;
	}

	return score;
}

int lock_tetfield(struct tetfield *tf)
{
	struct tetgrid *grid = tf->grid;
	blocks_row *blocks = grid->blocks;
	int num_lines_cleared = 0;
	int i;

	/* add blocks */
	xor_tetmino(&tf->mino, blocks);

	/* find completed lines */
	blocks += tf->mino.row;
	for (i=0; i < PIECE_HEIGHT; i++) {
		blocks_row b = blocks[i] + 1;
		if (!b) {
			num_lines_cleared++;
			/* init line clear */
			blocks[i] ^= LINE_CLEAR_MARK;
			grid->clearing++;
			grid->delay += BLOCK_CLEAR_DELAY;
		}
	}

	/* update score */
	tf->score += get_score_for_lines(tf, num_lines_cleared);

	/* update level */
	tf->pieces++;
	if (tf->pieces >= 25 && num_lines_cleared) {
		levelup(tf);
	}

	return num_lines_cleared;
}

File: lib/game/tetfield.js
import { TetGrid } from './tetgrid.js'
import { Piece, actions } from './piece.js'

/**
 * Tetromino playing field
 */
export class TetField {
  piece = new Piece()
  gravity = 20
  /** @type {typeof actions[number]} */
  lastAction = ''
  /** @type {Record<typeof actions[number], number>} */
  timeout = {
    '': 0,
    'rotate cw': 0,
    'rotate ccw': 0,
    'move right': 0,
    'move left': 0,
    harddrop: 0,
    softdrop: 0
  }

  constructor (grid = new TetGrid()) {
    this.grid = grid
  }

  /** prepare next tetromino */
  enter (piece = new Piece(), col = 0) {
    this.piece = piece
    piece.row = this.spawnRow + 1
    piece.col += col
    piece.falling = SPAWN_DELAY + 1

    this.state = TETFIELD_SPAWN

    // prevent last action from cancelling spawn delay
    this.stopRetryingAction()
  }

  get spawnRow () {
    return this.grid.blocks.length - this.piece.height
  }

  stopRetryingAction () {
    if (this.lastAction === HARDDROP) {
      this.lastAction = ''
      this.timeout[RETRY_ACTION] = 0
      this.timeout.harddrop += SPAWN_DELAY - AUTOREPEAT_FRAMES
    }
    if (this.lastAction) {
      this.lastAction = ''
      this.timeout[RETRY_ACTION] += SPAWN_DELAY - AUTOREPEAT_FRAMES
    }
  }

  makeMove (action = this.lastAction) {
    if (this.timeout[action] === 0 && this.piece.control(this.grid, action)) {
      if (!this.lastAction || !this.timeout[this.lastAction]) {
        this.lastAction = action
      }
      this.timeout[RETRY_ACTION] = AUTOREPEAT_FRAMES
      this.timeout[action] = AUTOREPEAT_FRAMES
      return action
    }
    return ''
  }

  /**
   * advance one frame
   *
   * @param {typeof actions[number]} action
   */
  run (action = '', out = {
    moved: action,
    dropped: 0
  }) {
    if (typeof action === 'number') {
      action = actions[action]
    }
    for (const a of actions) {
      if (this.timeout[a]) {
        this.timeout[a]--
      }
    }

    out.moved = ''
    out.dropped = 0

    switch (this.state) {
      case TETFIELD_SPAWN:
        this.updatePrespawn(action)
        if (this.piece.falling > 1 && this.grid.isClearing) {
          // stop spawn timer during line clear animation
          this.piece.falling++
        }
        out.dropped = this.piece.update(this.grid, this.gravity)
        if (!out.dropped) {
          if (!this.piece.isMovable && !this.grid.isClearing) {
            this.state = TETFIELD_TOP_OUT
          }
          return out
        }
        this.timeout[RETRY_ACTION] = 1
        this.state = TETFIELD_MOVE
        out.moved = this.updateMove(action)
        break
      case TETFIELD_MOVE:
        out.moved = this.updateMove(action)
        out.dropped = this.piece.update(this.grid, this.gravity)
        break
      case TETFIELD_PLACED:
        // accepting no more actions
        out.dropped = +this.piece.control(this.grid, HARDDROP)
        break
      case TETFIELD_TOP_OUT:
        return out
      default:
        break
    }

    // set timer to block actions for next tetromino during spawn delay
    if (this.timeout[RETRY_ACTION] === 1) {
      this.stopRetryingAction()
    }

    this.updateRemovableRows()

    if (this.piece.isMovable) {
      return out
    }

    this.state = TETFIELD_PLACED
    return out
  }

  /**
   * process moves before the piece is spawned
   * @param {typeof actions[number]} action
   */
  updatePrespawn (action) {
    if (!this.timeout[RETRY_ACTION]) {
      switch (action) {
        case 'rotate cw':
        case 'rotate ccw':
          this.lastAction = action
          break
        default:
          if (action && !this.timeout[action]) {
            this.piece.falling = 1
            this.timeout[action] = 1
          }
          break
      }
    }
  }

  /**
   * @param {typeof actions[number]} action
   */
  updateMove (action) {
    /** @type {typeof actions[number]} */
    let moved = ''

    if (this.timeout[action]) {
      if (this.timeout[RETRY_ACTION]) {
        moved = this.makeMove()
      }
      if (action) {
        this.lastAction = action
        this.timeout[RETRY_ACTION] = AUTOREPEAT_FRAMES
      }
    } else if (!(moved = this.makeMove(action))) {
      if (action) {
        this.lastAction = action
        this.timeout[RETRY_ACTION] = WALL_CHARGE_FRAMES
      }
    }

    return moved
  }

  /** check for collision with tetromino when removing cleared rows */
  updateRemovableRows () {
    if (this.grid.delay > 0) {
      return
    }

    let row = this.grid.getClearedRow()

    if (row > 0) {
      const grid2 = new TetGrid(this.grid.cols, this.grid.blocks.slice())

      // find row that causes collision with tetmino
      while (row > 0 && (grid2.shiftClearedBlocks(row) || this.piece.canMoveTo(grid2))) {
        row = this.grid.nextClearedRow(row)
      }

      this.grid.preventLineClears(row)
    }
  }

  get isLockable () {
    return this.state === TETFIELD_PLACED &&
      this.grid.isLockable(this.piece)
  }

  /**
   * lock tetromino into place and init line clear
   *
   * @return number of lines cleared
   */
  lock () {
    // add blocks
    this.grid.xorBlocks(this.piece)

    const numLinesCleared = this.grid.updateCompletedLines(this.piece.row)

    return numLinesCleared
  }
}

export const TETFIELD_SPAWN = 0
export const TETFIELD_MOVE = 1
export const TETFIELD_PLACED = 2
export const TETFIELD_TOP_OUT = 3

/**
 * Time constants are in frames assuming that the game will be run at this
 * speed
 */
export const TETFIELD_FPS = 60

/** auto repeat rate. This limits how fast an action can be repeated */
const AUTOREPEAT_FRAMES = 3

/** how many times a blocked move is retried */
const WALL_CHARGE_FRAMES = 4

/** time until appearance of next tetromino */
const SPAWN_DELAY = 25

/** `timeout[RETRY_ACTION]` is number of times to retry last action */
const RETRY_ACTION = ''

const HARDDROP = 'harddrop'
